pymtl3/datatypes/strategies.py

```diff
@@ -4,30 +4,114 @@
 ==========================================================================
 Hypothesis strategies for built-in data types.
 
-Author : Yanghui Ou
-  Date : June 16, 2019
+Author : Shunning Jiang, Yanghui Ou
+  Date : Dec 9, 2019
 """
 import hypothesis
 from hypothesis import strategies as st
 
-from .bits_import import mk_bits
+from .bits_import import Bits, mk_bits
+from .bitstructs import is_bitstruct_class
 
 #-------------------------------------------------------------------------
-# A generic bits strategy.
+# strategies.bits
 #-------------------------------------------------------------------------
+# Return the SearchStrategy for Bits_nbits with the support to limit
+# min/max value and setting signedness
 
-@st.composite
-def bits( draw, nbits, signed=False ):
-  if nbits == 1:
-    value = draw( st.booleans() )
-  elif not signed:
-    min_value = 0
-    max_value = 2**nbits - 1
-    value = draw( st.integers(min_value, max_value) )
-  else:
-    min_value = - ( 2 ** (nbits-1) )
-    max_value = 2**(nbits-1) - 1
-    value = draw( st.integers(min_value, max_value) )
-
+def bits( nbits, signed=False, min_value=None, max_value=None ):
   BitsN = mk_bits( nbits )
-  return BitsN( value )
+
+  if (min_value is not None or max_value is not None) and signed:
+    raise ValueError("bits strategy currently doesn't support setting "
+                     "signedness and min/max value at the same time")
+
+  if min_value is None:
+    min_value = (-(2**(nbits-1))) if signed else 0
+  if max_value is None:
+    max_value = (2**(nbits-1)-1)  if signed else (2**nbits - 1)
+
+  strat = st.booleans() if nbits == 1 else st.integers( min_value, max_value )
+
+  @st.composite
+  def strategy_bits( draw ):
+    return BitsN( draw( strat ) )
+
+  return strategy_bits() # RETURN A STRATEGY INSTEAD OF FUNCTION
+
+#-------------------------------------------------------------------------
+# strategies.bitslist
+#-------------------------------------------------------------------------
+# Return the SearchStrategy for a list of Bits with the support of
+# dictionary based min/max value limit
+
+def bitslist( types, limit_dict=None ):
```

Comment:
bitslists

Reply:
I think it should be bitslist ... it is a list of bits not multiple lists of bits?

Reply:
well in hypothesis it's also st.lists not list, because it's a strategy that generate different lists.

Reply:
Ah ha -- I stand corrected :) we should definitely follow the hypothesis naming conventions.